---
layout: page
title: Introduction to RNA-seq
excerpt: "Intro to RNA-seq"
date: 2018-03-27
startTime: 9.30
endTime: 17.00
city: sheffield
country: United Kingdom
venue: Pam Liversidge Design Studio 1 - D06
---

- Sheffield - 27th March 2017
- 09:30am - 5pm
- Pam Liversidge Design Studio 1 - D06

<iframe src="https://www.google.com/maps/embed?pb=!1m14!1m8!1m3!1d9519.181464571486!2d-1.4777067!3d53.3827108!3m2!1i1024!2i768!4f13.1!3m3!1m2!1s0x0%3A0x60e5580cdf19b137!2sPam+Liversidge+Building!5e0!3m2!1sen!2suk!4v1510862811609" width="600" height="450" frameborder="0" style="border:0" allowfullscreen></iframe>

## Overview

High-throughput RNA-sequencing is now the standard technique for quantifying transcript abundance in a biological sample of interest. In this course we will describe the processes that take place once you submit a library for RNA sequencing, and what data you should expect to receive from the Bioinformatics Core.

We will describe the steps involved to go from sequencing library to a list of genes that show statistically significant differences between your biological conditions of interest. Practical sessions will use the user-friendly Galaxy interface (https://usegalaxy.org/) to demonstrate tasks such as *alignment*, *quality control* and *assessing differential expression*. We will also showcase some web sites you can use for enrichment and pathways analysis.

Please note that the course will not cover the analysis of RNA-seq data in R

## Who should attend this course?

Researchers in life sciences who want to get an appreciation for the computational steps involved in RNA-seq analysis, but not neccesarily wishing to execute the pipeline for themselves. 

## Objectives:- After this course you should be able to:

- Appreciate some of the issues that can arise when designing an RNA-seq experiment 
- Recall the key pipeline steps in an RNA-seq analysis
- Understand the contents of fastq and bam file format for RNA-seq data
- Perform a gene set enrichment analysis using available online tools

## Aims:- During this course you will learn about:

- Basic principles of Experimental design for RNA-seq experiments
- The steps in a best-practice pipeline for RNA-seq analysis
- The theory behind popular methods for pathways and gene set enrichment analysis
- Executing basic Bioinformatics tools using Galaxy

## Software installation



## Instructors

- Mark Dunning, Bioinformatics Core Director
- Emily Chambers, Bioinformatics Core Analyst

## Course Data

## Timetable (provisional)

- 09:30 - 10:30 Introduction to RNA-seq and experimental design - Lecture
- 10:30 - 12:30 [Hands-on tutorial on RNA-seq data formats, alignment and differential expression](http://sbc.shef.ac.uk/rna-seq-in-galaxy/)
LUNCH
- 13:30 - 14:30 Tutorial continued
- 14:30 - 17:00 [Hands-on tutorial on gene set enrichment](https://sheffield-bioinformatics-core.github.io/enrichment-analysis/enrichment.nb.html
)


## Registration 

Registration is open. [Click here](https://onlineshop.shef.ac.uk/conferences-events/faculty-of-medicine-dentistry-and-health/neuroscience/introduction-to-rnaseq) to sign-up

